[CALCITE-5008] Ignore synthetic and static methods in MetadataDef

rpuch opened the pull request:
According to https://www.jacoco.org/jacoco/trunk/doc/faq.html , Jacoco
adds a synthetic static method called $jacocoInit() during
instrumentation.

MetadataDef constructor breaks if it encounters such method on
the class given to it because it does not expect it. But it looks like
it simply should not consider synthetic methods (as well as static
methods) at all.

rpuch commented:
A test has failed.

```
FAILURE   0.1sec, org.apache.calcite.test.DruidAdapter2IT > testAggOverStringToLong()
    java.lang.IllegalStateException: Unable to create connection: connectionFactory = org.apache.calcite.test.ConnectionFactories$MapConnectionFactory@f20d5180
        at org.apache.calcite.test.CalciteAssert$AssertQuery.createConnection(CalciteAssert.java:1375)
        at org.apache.calcite.test.CalciteAssert$AssertQuery.withConnection(CalciteAssert.java:1383)
        at org.apache.calcite.test.CalciteAssert$AssertQuery.queryContains(CalciteAssert.java:1699)
        at org.apache.calcite.test.DruidAdapter2IT.testAggOverStringToLong(DruidAdapter2IT.java:3463)
        Caused by: java.sql.SQLException: No suitable driver found for jdbc:calcite:
            at java.sql.DriverManager.getConnection(DriverManager.java:689)
            at java.sql.DriverManager.getConnection(DriverManager.java:208)
            at org.apache.calcite.test.ConnectionFactories$MapConnectionFactory.createConnection(ConnectionFactories.java:115)
            at org.apache.calcite.test.CalciteAssert$AssertQuery.createConnection(CalciteAssert.java:1373)
            ... 3 more
            Next exception 1: [CIRCULAR REFERENCE SQLException]
```

It does not seem to be related to my changes. Is it possible to re-run just the failed sub-task? I did not find one, maybe I'm not looking at the correct place, or maybe I just don't have enough privileges.

rubenada commented on core/src/main/java/org/apache/calcite/rel/metadata/MetadataDef.java:
```diff
@@ -41,7 +42,10 @@ private MetadataDef(Class<M> metadataClass,
     this.methods = ImmutableList.copyOf(methods);
     final Method[] handlerMethods = Arrays.stream(handlerClass.getDeclaredMethods())
-        .filter(m -> !m.getName().equals("getDef")).toArray(i -> new Method[i]);
+        .filter(m -> !m.getName().equals("getDef"))
```
@jacques-n it seems this filter was originally introduced via CALCITE-4929 (see [4ff5fa8](https://github.com/apache/calcite/commit/4ff5fa89c180ebc30d8fb324c2a50b0e9797b9ca) ), what do you think about the current patch?
Also, I noticed that in  CALCITE-4929 a similar filter was introduced in [RelMetadataHandlerGeneratorUtil.java](https://github.com/apache/calcite/commit/4ff5fa89c180ebc30d8fb324c2a50b0e9797b9ca#diff-b5bac3d2a905799f710aef8e5fd02dd659dc9e6e2c8e7b05ed0c34022025b8e1) , if we apply this patch, should we apply the change in both files too?

jacques-n replied:
Yes, we should apply the same logic in both places. I was thinking that we should probably extract this to a utility method as opposed to continue duplication. As part of this, let's also add specific units tests which verify the utility method. You up for that @rpuch?

rpuch replied:
@jacques-n sure, I'll do it. Thank you guys for your suggestions.

rpuch replied:
Done

rpuch commented:
@rubenada @jacques-n I extracted the code for finding handler methods and also added tests concerning synthetic methods. Could you please take a look?

jacques-n commented on core/src/test/java/org/apache/calcite/rel/metadata/BlankMetadataHandler.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * A blank {@link MetadataHandler} that is used as a base for adding a synthetic method.
+ */
+interface BlankMetadataHandler extends MetadataHandler<TestMetadata> {
```
Can you move the test interfaces into the test class files?

rpuch replied:
Moved `BlankMetadataHandler` into a test class.

`TestMetadata` and `MetadataHandlerWithStaticMethod` are used by two distinct test classes, so they cannot be 'internalized' without fusing the test classes together. If your suggestion is to still fuse them, please elaborate what test structure you would prefer to see.

jacques-n commented on core/src/main/java/org/apache/calcite/rel/metadata/MetadataHandlerMethods.java:
```diff
@@ -0,0 +1,45 @@
+   * @param handlerClass the handler class to inspect
+   * @return handler methods
+   */
+  public static Method[] handlerMethods(Class<? extends MetadataHandler<?>> handlerClass) {
```
How about we put this as a static method on MetadataHandler instead of a separate class?

rpuch replied:
Moved the method to `MetadataHandler` class

jacques-n reviewed (commented):
Generally looks good. Would be great if you could collapse a bunch of the small test files to static inner classes/interfaces (within the test classes).

rpuch commented:
@rubenada @jacques-n guys is anything still missing?

jacques-n commented on build.gradle.kts:
```diff
@@ -893,3 +893,7 @@ allprojects {
     }
 }
+java {
+    sourceCompatibility = JavaVersion.VERSION_17
```
We shouldn't be changing this as part of this patch. (I didn't notice it before.)

The rest looks good.

rpuch replied:
Oops, good catch. For some reason, it's already second time when this change was added. First time I noticed and reverted it, second time I didn't. I wonder whether it's added by an IDE (I use IntelliJ IDEA) or by some build script.

Anyway, I removed the change. Could we please proceed?

jacques-n reviewed (approved):
LGTM.

rpuch commented:
Thank you for your review!

jacques-n commented:
Thanks for the contribution @rpuch!

rpuch commented:
Thank you guys!
